src/linear_algebra/sesquilinear_form.lean

```diff
@@ -33,68 +34,107 @@ Sesquilinear form,
+end comm_ring
+section field
 
+variables [field K] [add_comm_group V] [module K V]
+  [field K₁] [add_comm_group V₁] [module K₁ V₁]
+  [field K₂] [add_comm_group V₂] [module K₂ V₂]
+  {I₁ : K₁ →+* K} {I₂ : K₂ →+* K} {I₁' : K₁ →+* K}
+  {J₁ : K →+* K} {J₂ : K →+* K}
 
-lemma ortho_smul_left {x y} {a : R} (ha : a ≠ 0) : (is_ortho B x y) ↔ (is_ortho B (a • x) y) :=
+-- todo: this also holds for [comm_ring R] [is_domain R] when J₁ is invertible
+lemma ortho_smul_left {B : V₁ →ₛₗ[I₁] V₂ →ₛₗ[I₂] K} {x y} {a : K₁} (ha : a ≠ 0) :
+  (is_ortho B x y) ↔ (is_ortho B (a • x) y) :=
 begin
   dunfold is_ortho,
   split; intro H,
-  { rw [map_smul, smul_apply, H, smul_zero] },
-  { rw [map_smul, smul_apply, smul_eq_zero] at H,
+  { rw [map_smulₛₗ₂, H, smul_zero]},
+  { rw [map_smulₛₗ₂, smul_eq_zero] at H,
     cases H,
-    { trivial },
+    { rw I₁.map_eq_zero at H, trivial },
     { exact H }}
 end
 
-lemma ortho_smul_right {x y} {a : R} {ha : a ≠ 0} : (is_ortho B x y) ↔ (is_ortho B x (a • y)) :=
+-- todo: this also holds for [comm_ring R] [is_domain R] when J₂ is invertible
+lemma ortho_smul_right {B : V₁ →ₛₗ[I₁] V₂ →ₛₗ[I₂] K} {x y} {a : K₂} {ha : a ≠ 0} :
+(is_ortho B x y) ↔ (is_ortho B x (a • y)) :=
 begin
   dunfold is_ortho,
   split; intro H,
   { rw [map_smulₛₗ, H, smul_zero] },
   { rw [map_smulₛₗ, smul_eq_zero] at H,
     cases H,
-    { rw [ring_equiv.to_ring_hom_eq_coe, ring_equiv.coe_to_ring_hom] at H,
+    { simp at H,
       exfalso,
-      exact ha (I.map_eq_zero_iff.mp H) },
+      exact ha H },
     { exact H }}
 end
 
-end is_domain
+/-- A set of orthogonal vectors `v` with respect to some sesquilinear form `B` is linearly
+  independent if for all `i`, `B (v i) (v i) ≠ 0`. -/
+lemma linear_independent_of_is_Ortho
+  {n : Type*} {B : V₁ →ₛₗ[I₁] V₁ →ₛₗ[I₁'] K} {v : n → V₁}
+  (hv₁ : B.is_Ortho v) (hv₂ : ∀ i, ¬ B.is_ortho (v i) (v i)) :
+  linear_independent K₁ v :=
+begin
+  classical,
+  rw linear_independent_iff',
+  intros s w hs i hi,
+  have : B (s.sum $ λ (i : n), w i • v i) (v i) = 0,
+  { rw [hs, map_zero, zero_apply] },
+  have hsum : s.sum (λ (j : n), I₁(w j) * B (v j) (v i)) = I₁(w i) * B (v i) (v i),
+  { apply finset.sum_eq_single_of_mem i hi,
+    intros j hj hij,
+    rw [is_Ortho_def.1 hv₁ _ _ hij, mul_zero], },
+  simp_rw [B.map_sum₂, map_smulₛₗ₂, smul_eq_mul, hsum] at this,
+  apply I₁.map_eq_zero.mp,
+  exact eq_zero_of_ne_zero_of_mul_right_eq_zero (hv₂ i) this,
+end
+
+end field
 
-variables {R : Type*} {M : Type*} [comm_ring R] [add_comm_group M] [module R M]
-  {I : R →+* R}
-  {B : M →ₗ[R] M →ₛₗ[I] R}
+variables [comm_ring R] [add_comm_group M] [module R M]
+  [comm_ring R₁] [add_comm_group M₁] [module R₁ M₁]
+  {I : R →+* R} {I₁ : R₁ →+* R} {I₂ : R₁ →+* R}
+  {B : M₁ →ₛₗ[I₁] M₁ →ₛₗ[I₂] R}
+  {B' : M →ₗ[R] M →ₛₗ[I] R}
 
 /-- The proposition that a sesquilinear form is reflexive -/
-def is_refl (B : M →ₗ[R] M →ₛₗ[I] R) : Prop :=
+def is_refl (B : M₁ →ₛₗ[I₁] M₁ →ₛₗ[I₂] R) : Prop :=
   ∀ (x y), B x y = 0 → B y x = 0
```

Comment:
@eric-wieser does it make sense to turn these types of defs into classes?

Reply:
I could certainly see this being used as a class, a bit like `filter.ne_bot`. It's a bit hard to tell if that would be better without having more code about this.

Reply:
yes that is a good point.